kong/resty/dns_client/utils.lua

```diff
@@ -0,0 +1,270 @@
+  if not lines then
+    log(NOTICE, "Invalid hosts file: ", err)
+    return DEFAULT_HOSTS
+  end
+
+  local hosts = {}
+
+  for _, line in ipairs(lines) do
+    -- Remove leading/trailing whitespaces and split by whitespace
+    local parts = {}
+    for part in line:gmatch("%S+") do
+      if part:sub(1, 1) == '#' then
+        break
+      end
+      table_insert(parts, part:lower())
+    end
+
+    -- Check if the line contains an IP address followed by hostnames
+    if #parts >= 2 then
+      local ip, _, family = _M.parse_hostname(parts[1])
+      if family ~= "name" then    -- ipv4/ipv6
+        for i = 2, #parts do
+          local host = parts[i]
+          local v = hosts[host]
+          if not v then
+            v = {}
+            hosts[host] = v
+          end
+          v[family] = v[family] or ip -- prefer to use the first ip
+        end
+      end
+    end
+  end
+
+  if not hosts.localhost then
+    hosts.localhost = LOCALHOST
+  end
+
+  return hosts
+end
+
+
+-- TODO: need to rewrite it instead of calling parseResolvConf
+function _M.parse_resolv_conf(path, enable_ipv6)
+  local resolv, err = utils.parseResolvConf(path or DEFAULT_RESOLV_CONF)
+  if not resolv then
+    return nil, err
+  end
+
+  resolv = utils.applyEnv(resolv)
+  resolv.options = resolv.options or {}
+  resolv.ndots = resolv.options.ndots or 1
+  resolv.search = resolv.search or (resolv.domain and { resolv.domain })
+
+  -- check if timeout is 0s
+  if resolv.options.timeout and resolv.options.timeout <= 0 then
+    log(NOTICE, "A non-positive timeout of ", resolv.options.timeout,
+                "s is configured in resolv.conf. Setting it to 2000ms.")
+    resolv.options.timeout = 2000 -- 2000ms is lua-resty-dns default
+  end
+
+  -- remove special domain like "."
+  if resolv.search then
+    for i = #resolv.search, 1, -1 do
+      if resolv.search[i] == "." then
+        table_remove(resolv.search, i)
+      end
+    end
+  end
+
+  -- nameservers
+  if resolv.nameserver then
+    local nameservers = {}
+
+    for _, address in ipairs(resolv.nameserver) do
+      local ip, port, t = utils.parseHostname(address)
+      if t == "ipv4" or
+        (t == "ipv6" and not ip:find([[%]], nil, true) and enable_ipv6)
+      then
+        table_insert(nameservers, port and { ip, port } or ip)
+      end
+    end
+
+    resolv.nameservers = nameservers
+  end
+  return resolv
+end
+
+
+function _M.is_fqdn(name, ndots)
+  local _, dot_count = name:gsub("%.", "")
+  return (dot_count >= ndots) or (name:sub(-1) == ".")
+end
+
+
+-- construct names from resolv options: search, ndots and domain
+function _M.search_names(name, resolv, hosts)
+  if not resolv.search or _M.is_fqdn(name, resolv.ndots) then
+    return { name }
+  end
+
+  local names = {}
+  for _, suffix in ipairs(resolv.search) do
+    table_insert(names, name .. "." .. suffix)
+  end
+
+  if hosts and hosts[name] then
+    table_insert(names, 1, name)
+  else
+    table_insert(names, name)
+  end
```

Comment:
iiuc `hosts` contains the `/etc/hosts` listed hostnames. Why are we preferring them when searching? they are inserted at the start, where other search options go to the end of the list.

Reply:
ok, nvm. Since the names to search for are fully qualified, that is safe. But then why do we insert the name at the start of the table?
Can't we just do this check before looping over the search table? and if it matches return a table with only that entry?

Stated otherwise; if the name matches a name in `/etc/hosts`, then why do we need the search domains as well?

Reply:
Could we add some comments to explain this code block?

Reply:
> Can't we just do this check before looping over the search table? and if it matches return a table with only that entry?

1. As I wrote this logic, I tried to align it with the original DNS library, I found that it also searches for other names generated by `search` option in `resolv.conf`, such as an example test case [here](https://github.com/Kong/kong/blob/master/spec/01-unit/21-dns-client/02-client_spec.lua#L555). 
2. Additionally I think there is a small probability that the hosts' static IP addresses may be evicted from the LRU cache and shared dict, if this scenario happened, we could try more names to give it a chance to get the right records.

The final solution might involve ensuring that hosts' static IP addresses are not evicted from the cache, or implementing specific logic to check a hostnames before looking them up in the cache.

Reply:
Agree, I will add some comments here.

Reply:
add comments here: https://github.com/Kong/kong/pull/12305/commits/b87cc14ab796307cf88d07c7aed08475a9bb0497

Reply:
It should not be possible to miss the host entry because of an LRU eviction. In the old DNS client we had a fix for this, iirc on every cache lookup failure, also check the hosts list. This behaviour should be preserved to prevent regressions.

Reply:
> In the old DNS client we had a fix for this, iirc on every cache lookup failure, also check the hosts list.

I reviewed the original DNS client library, not found this logic.

and I also wrote a simple test case to check it, if host entries are evicted from LRUcache, the original DNS client would query it from the DNS server. And if it failed to get result, it returned an error.

```
    describe("inject localhost: #ttt", function()

      it("if absent #ttt", function()
        local result, err, record
        result, err = assert(client.init({
            nameservers = { {"8.8.8.8", 53} },
            resolvConf = {},
            hosts = {},
          }))
        assert.is.True(result)
        assert.is.Nil(err)
        record = client.getcache():get("28:localhost")
        assert.equal("[::1]", record[1].address)
        record = client.getcache():get("1:localhost")
        assert.equal("127.0.0.1", record[1].address)

        do  -- evict localhost keys
          client.getcache():delete("28:localhost")
          client.getcache():delete("1:localhost")
          assert(client.getcache():get("1:localhost") == nil)
          assert(client.getcache():get("28:localhost") == nil)
        end

        local res, err, tries = client.resolve("localhost")
        assert.same(err, "")
        assert.same(res, {})

        error(1)
      end)
```

Reply:
But never mind, I will update some logic to fix the hosts eviction scenario in this PR

Reply:
resolved in https://github.com/Kong/kong/pull/12305/commits/2dddd544a6cecf966d4df36ba169e0711c845533.

Reply:
hmmm weird, we had that case from a user/customer and we did fix it iirc. Thx for fixing it!